addons/l10n_my_edi_extended/models/account_edi_xml_ubl_my.py

```diff
@@ -50,14 +65,22 @@ def _export_invoice_constraints(self, invoice, vals):
     @api.model
     def _l10n_my_edi_get_document_type_code(self, invoice):
         """ Override the super method to include self billed documents. """
-        # OVERRIDE 'l10n_my_edy'
+        # OVERRIDE 'l10n_my_edi'
         super()._l10n_my_edi_get_document_type_code(invoice)
 
         if 'debit_origin_id' in self.env['account.move']._fields and invoice.debit_origin_id:
             code = '03' if invoice.move_type == 'out_invoice' else '13'
             return code, invoice.debit_origin_id
         elif invoice.move_type in ('out_refund', 'in_refund'):
-            code = '02' if invoice.move_type == 'out_refund' else '12'
+            # We consider a credit note a refund if it is paid and fully reconciled with a payment or bank transaction.
+            is_paid = invoice.payment_state in ('paid', 'in_payment') and invoice.currency_id.compare_amounts(invoice.amount_residual, 0) == 0
+            reconcile_lines = invoice.line_ids.filtered(lambda x: x.account_id.reconcile or x.account_id.account_type in ('asset_cash', 'liability_credit_card'))
+            reconciled_entry = reconcile_lines.full_reconcile_id.reconciled_line_ids.move_id.filtered(lambda x: x.is_entry())
+            if not is_paid or not reconcile_lines.full_reconcile_id or not reconciled_entry:
+                code = '02' if invoice.move_type == 'out_refund' else '12'
+            else:
+                code = '04' if invoice.move_type == 'out_refund' else '14'
```

Comment:
What about:
```py
payment_terms = invoice.line_ids.filtered(lambda aml: aml.display_type == 'payment_term')
counterpart_amls = payment_terms.matched_debit_ids.debit_move_id + payment_terms.matched_credit_ids.credit_move_ids
counterpart_move_type = 'out_invoice' if invoice.move_type == 'out_refund' else 'out_refund'
has_payments = bool(counterpart_amls.filtered(lambda aml: aml.move_id.move_type != counterpart_move_type)
is_paid = invoice.payment_state == 'paid'
if is_paid and has_payments:
    code = '04' if invoice.move_type == 'out_refund' else '14'
else:
    code = '02' if invoice.move_type == 'out_refund' else '12'
```

Reply:
At first glance it seems ok if you replace the is_paid line by
`is_paid = invoice.payment_state == invoice._get_invoice_in_payment_state()`

I'll check with Doh to get it tested more, thanks!

Reply:
Update: after testing it works in most use cases, but there is an issue with refund for self billed invoice.
Will let you two know once corrected, likely tomorrow.